### ISSUE TITLE
Level 9 race condition makes solution very hard to find

### ISSUE THREAD

**Issue by DracoDynasty** [andybrewer/operation-go#1]
On level 9 the race condition is VERY precise and affected by more things that what the solution leads you to believe.

When naming your variable, the exact number of letters you use will affect the end order. 
Calling it "m" for example resulted in a wrong order for the printed code.

Another example: when using the exact name given in the solution, but not adding a blank line between the newly defined variable and the start of setButton will also result in the wrong order.

I'm not sure what exactly causes this, and while I didn't get the solution by myself I was baffled to see that it didn't work until it was a perfect copy/paste of the solution in the repository.

**Comment by andybrewer** [comment]
Wow, good catch! I was wondering if it was the playground, but even on my local machine running `1.13`, adding white space lines varies the output. This game has been in the wild for a few years now, so I'm wondering if something changed with `mutex` or if this was always a bug. Any thoughts on how to maintain the theme of the level but increase the reproducibility?

**Comment by DracoDynasty** [comment]
TL;DR the more I tried to remove the surbroutin-induced randomness the less I succeeded. Sadly I didn't find a solution.


On the other hand there was something else that bothered me with that solution and the expected output.

To me, it didn't make sense that the output should be anything other than the original order from the buttons var once you added mutex. Since the sleep was then inside a lock, it shouldn't matter anymore and the order should be kept.

However, the accepted solution started with purple, the last color from the list. So something else must have been at hand.


I decided to test some more to see how the program responded.

First, I got back to the stage in the page and tried again to copy-paste the solution to make sure it always was accepted at all tries... Thankfully I didn't have to reset it multiple times as the first try failed, then the second with the exact same code succeeded.

Because of that and the fact that previously I have been able to get it first try with copypasting, I thought that maybe the go subroutines didn't launch in the same exact order at each run...

Next I took the code to run it locally and try to test some more. I modified the code so that it didn't sleep anymore and instead printed x and the color of the button. Basically:
```
var mutex sync.Mutex
func setButton(x int, sequence *[]Button) {
    mutex.Lock()
    defer mutex.Unlock()
    newButton := Button{buttons[x]}
    println("    N°", x, ":", newButton.color)
    *sequence = append(*sequence, newButton)
}
```

Again, I observed the same outcome. Each run had the potential to have a different order, meaning the order the go subroutines were running their respective appends was not deterministic, even without any random sleep time.

I then moved the lock in addButton to see if we could do away with setButton as a solution to our problem, but I still observed different orders of output at this point. Basically, the race condition is so fast that even running a lock as the first instruction in the subroutine does not guarantee the order.
I then tried to create the mutex in main() and pass a ref to it in addButton and still ended up with randomness, which I find crazy.

At this point I'm not sure at all this can be solved by keeping that same premise.

**Comment by andybrewer** [comment]
This is puzzling and unfortunate. I was able to replicate the issues you described. I wonder if Go's playground randomness or mutexes have changed since Op Go was originally released in 2015.  I think unfortunately a new level is in order.

**Comment by pkitszel** [comment]
Hi, I find the whole game pretty interesting (ok, I did not yet pass lvl 10 as of this writing :P).

I also would expect to see the same sequence as OP as correct one.

I would guess that `purple` goroutine is the one running after spawning because main simply has no other job to do but switch to last one (this is ofc implementation dependend, but I have spotted similar behaviour multiple times).

Mutex indeed fixes all race related problems here, except the one about order ;)

To keep this task as close to current one, I would suggest to just unfreeze more code to edit for user (to keep it easy), or just left it as is, but change correct answer (but then code for it is quite mangled (find my solution below (which yields always the same output color sequence, without races)).

Finally, If you want to keep everythig as-is, my solution to pass it needs to replace `(*sequence)[x]` by `(*sequence)[(x+1)%5]`

```go
var mtx sync.Mutex

func fixMe(sequence *[]Button) {
    if (len(*sequence) == 10) {
        *sequence = (*sequence)[:5]
    }
}

func setButton(x int, sequence *[]Button) {
    mtx.Lock()
    defer mtx.Unlock()
    defer fixMe(sequence)
    if (len(*sequence) == 0) {
        *sequence = make([]Button, 5, 10)
    }
    (*sequence)[x] = Button{buttons[x]}
	
	newButton := Button{buttons[x]}
	time.Sleep(time.Duration(rand.Intn(10)) * time.Millisecond)
	*sequence = append(*sequence, newButton)
}

**Comment by andybrewer** [comment]
The game is quite old at this point and will probably need to be updated when Go 2 comes out soon, so I've decided to simply reference this issue in the level's code: https://github.com/andybrewer/operation-go/blob/master/levels/level_9/level_9.go

For anyone stuck on this level, you can copy/paste this solution (no extra whitespace or line breaks, per the comments above). Thanks @DracoDynasty and @pkitszel for your help on this!

```
var mutex sync.Mutex

func setButton(x int, sequence *[]Button) {
    mutex.Lock()
    defer mutex.Unlock()
```

**Comment by oPOCCOMAXAo** [comment]
Just pass validation, without real solution:
```go
func setButton(_ int, sequence *[]Button) {
    *sequence = []Button{
        {"purple"},
        {"red"},
        {"blue"},
        {"green"},
        {"yellow"},
    }
}

func unused(){
    var x int
    var sequence *[]Button
```
